hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java

```diff
@@ -1,54 +1,50 @@
-    private final SchemaRepository<Schema> avroSchemaRepository;
 
     @Inject
     public MessageContentWrapper(JsonMessageContentWrapper jsonMessageContentWrapper,
-                                 AvroMessageContentWrapper avroMessageContentWrapper,
-                                 SchemaRepository<Schema> avroSchemaRepository) {
+                                 AvroMessageContentWrapper avroMessageContentWrapper) {
         this.jsonMessageContentWrapper = jsonMessageContentWrapper;
         this.avroMessageContentWrapper = avroMessageContentWrapper;
-        this.avroSchemaRepository = avroSchemaRepository;
     }
 
-    public UnwrappedMessageContent unwrap(byte[] data, Topic topic) {
-        return unwrap(data, topic, topic.getContentType());
+    public UnwrappedMessageContent unwrapJson(byte[] data) {
+        return jsonMessageContentWrapper.unwrapContent(data);
     }
 
-    public UnwrappedMessageContent unwrap(byte[] data, Topic topic, ContentType contentType) {
-        if (contentType == ContentType.JSON) {
-            return jsonMessageContentWrapper.unwrapContent(data);
-        } else if (contentType == ContentType.AVRO) {
-            return avroMessageContentWrapper.unwrapContent(data, avroSchemaRepository.getSchema(topic));
-        }
-
-        throw new UnsupportedContentTypeException(topic);
+    public UnwrappedMessageContent unwrapAvro(byte[] data, Topic topic, Function<Optional<SchemaVersion>, CompiledSchema<Schema>> schemaProvider) {
+        return topic.isSchemaVersionAwareSerializationEnabled() ?
+                deserialize(data, schemaProvider) : avroMessageContentWrapper.unwrapContent(data, schemaProvider.apply(empty()));
     }
 
-    public byte[] wrap(byte[] data, String id, long timestamp, Topic topic, Map<String, String> externalMetadata) {
-        return wrap(data, id, timestamp, topic, topic.getContentType(), externalMetadata);
+    private UnwrappedMessageContent deserialize(byte[] data, Function<Optional<SchemaVersion>, CompiledSchema<Schema>> schemaProvider) {
+        SchemaAwareSerializer.DeserializedPayload payload = SchemaAwareSerializer.deserialize(data);
+        return avroMessageContentWrapper.unwrapContent(payload.getPayload(), schemaProvider.apply(of(payload.getSchemaVersion())));
     }
 
-    public byte[] wrap(byte[] data, String id, long timestamp, Topic topic, ContentType contentType, Map<String, String> externalMetadata) {
-        if (contentType == ContentType.JSON) {
-            return jsonMessageContentWrapper.wrapContent(data, id, timestamp, externalMetadata);
-        } else if (contentType == ContentType.AVRO) {
-            return avroMessageContentWrapper.wrapContent(data, id, timestamp, avroSchemaRepository.getSchema(topic), externalMetadata);
-        }
-
-        throw new UnsupportedContentTypeException(topic);
+    public byte[] wrapAvro(byte[] data, String id, long timestamp, Topic topic, CompiledSchema<Schema> schema, Map<String, String> externalMetadata) {
```

Comment:
Unfortunately this is very difficult because this class is used in both consumer and frontend modules which have very different definitions of a Message. I thought about introducing some 3rd implementation just to manage number of arguments here but we would have 1 argument more instead of 3. I'm not so sure if adding another abstraction for a message makes things easier to be honest...